### PR TITLE
Add ability to swap items in storage

### PR DESCRIPTION
**Pull request by dearmochi**
## What Does This PR Do
- Adds ability to swap items in storage through drag and drop
- Fixes "the the" message when putting item in storage
- Cleans code where possible

## Why It's Good For The Game

Makes inventory management less of a pain

## Images of changes

T0ejj56QZR

## Changelog
:cl:
add: Add ability to swap items in storage (backpack, boxes, etc.)
spellcheck: Fix "the the" when inserting an item to storage
/:cl:

### CODE REVIEW

**Comment by Mitchs98** [comment]
This is beautiful.

File: code/_onclick/drag_drop.dm [hunk 1]
```diff
@@ -7,10 +7,11 @@
 */
 /atom/MouseDrop(atom/over)
 	if(!usr || !over) return
-	if(!Adjacent(usr) || !over.Adjacent(usr)) return // should stop you from dragging through windows
+	if(!istype(over, /obj/screen))
+		if(!Adjacent(usr) || !over.Adjacent(usr)) return // should stop you from dragging through windows
```
Comment (SteelSlayer) [inline code/_onclick/drag_drop.dm:11]: I know this was already like this but, can you put the return on its own line?

Comment (dearmochi) [reply code/_onclick/drag_drop.dm:11]: Aye, done

**Comment by dearmochi** [comment]
Spent the better part of an hour writing a snowflake solution for numbered storages when I realized the one for non-numbered storages worked perfectly fine and amounted to pretty much the same thing, except much less verbose. That's why I force-pushed. Here's the comment I'd wrote anyway explaining the process:

> Basically the idea of this is that we're going to swap the FIRST occurrences of both items we're trying to swap.
Example: A A B C C -> C A B A C
This exploits the fact that orient2hud orders the displayed stacks by the position of the first item of each stack.
While it doesn't result in a neatly grouped list, it maintains the order of order items and is relatively inexpensive

File: code/_onclick/hud/screen_objects.dm [hunk 2]
```diff
@@ -169,7 +169,7 @@
 /obj/screen/storage/Click(location, control, params)
 	if(world.time <= usr.next_move)
 		return 1
-	if(usr.stat || usr.paralysis || usr.stunned || usr.IsWeakened())
+	if(usr.incapacitated())
```
Comment (Fox-McCloud) [inline code/_onclick/hud/screen_objects.dm:172]: Change in functionality; this needs to ignore lying down or restraints.

Comment (dearmochi) [reply code/_onclick/hud/screen_objects.dm:172]: Whoops, fixed

File: code/_onclick/drag_drop.dm [hunk 3]
```diff
@@ -6,11 +6,14 @@
 	almost anything into a trash can.
 */
 /atom/MouseDrop(atom/over)
-	if(!usr || !over) return
-	if(!Adjacent(usr) || !over.Adjacent(usr)) return // should stop you from dragging through windows
+	if(!usr || !over)
+		return
+	if(!(istype(over, /obj/screen) || (src.loc && src.loc == over.loc)))
```
Comment (farie82) [inline code/_onclick/drag_drop.dm:11]: `src` is implied

File: code/_onclick/hud/screen_objects.dm [hunk 4]
```diff
@@ -179,6 +179,39 @@
 			master.attackby(I, usr, params)
 	return 1
 
+/obj/screen/storage/MouseDrop_T(obj/item/I, mob/user)
+	if(!user || !istype(I) || usr.incapacitated(ignore_restraints = TRUE, ignore_lying = TRUE) || istype(user.loc, /obj/mecha) || !master)
```
Comment (farie82) [inline code/_onclick/hud/screen_objects.dm:183]: `user` instead of `usr` here

File: code/game/objects/items/weapons/storage/storage.dm [hunk 5]
```diff
@@ -21,10 +21,10 @@
 	var/display_contents_with_number	//Set this to make the storage item group contents of the same type and display them as a number.
 	var/allow_quick_empty	//Set this variable to allow the object to have the 'empty' verb, which dumps all the contents on the floor.
 	var/allow_quick_gather	//Set this variable to allow the object to have the 'toggle mode' verb, which quickly collects all items from a tile.
-	var/collection_mode = 1;  //0 = pick one at a time, 1 = pick all on tile
+	var/collection_mode = TRUE  //0 = pick one at a time, 1 = pick all on tile
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:24]: Best to also change the comment

File: code/game/objects/items/weapons/storage/storage.dm [hunk 6]
```diff
@@ -126,28 +125,27 @@
...
 	user.client.screen -= src.contents
 	if(user.s_active == src)
 		user.s_active = null
 	return
 
-/obj/item/storage/proc/open(mob/user as mob)
+/obj/item/storage/proc/open(mob/user)
 	if(src.use_sound)
-		playsound(src.loc, src.use_sound, 50, 1, -5)
+		playsound(src.loc, src.use_sound, 50, TRUE, -5)
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:141]: src implied

File: code/game/objects/items/weapons/storage/storage.dm [hunk 7]
```diff
@@ -204,68 +202,67 @@
 	var/obj/item/sample_object
 	var/number
 
-	New(obj/item/sample as obj)
+	New(obj/item/sample)
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:205]: Gross way to declare procs.
Best to write this as `/datum/numbered_display/New(...)`

File: code/game/objects/items/weapons/storage/storage.dm [hunk 8]
```diff
@@ -204,68 +202,67 @@
...
 
 	//var/mob/living/carbon/human/H = user
 	var/row_num = 0
-	var/col_count = min(7,storage_slots) -1
+	var/col_count = min(7, storage_slots) - 1
 	if(adjusted_contents > 7)
-		row_num = round((adjusted_contents-1) / 7) // 7 is the maximum allowed width.
-	src.standard_orient_objs(row_num, col_count, numbered_contents)
+		row_num = round((adjusted_contents - 1) / 7) // 7 is the maximum allowed width.
+	src.standard_orient_objs(row_num, col_count, display_contents)
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:236]: src implied

File: code/game/objects/items/weapons/storage/storage.dm [hunk 9]
```diff
@@ -323,15 +320,15 @@
...
 	W.in_inventory = TRUE
 	update_icon()
-	return 1
+	return TRUE
 
 //Call this proc to handle the removal of an item from the storage item. The item will be moved to the atom sent as new_target
-/obj/item/storage/proc/remove_from_storage(obj/item/W as obj, atom/new_location)
-	if(!istype(W)) return 0
+/obj/item/storage/proc/remove_from_storage(obj/item/W, atom/new_location)
+	if(!istype(W)) return FALSE
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:327]: Best to put this on the next line. Makes it more readable

File: code/game/objects/items/weapons/storage/storage.dm [hunk 10]
```diff
@@ -392,8 +389,8 @@
 
 	handle_item_insertion(I)
 
-/obj/item/storage/attack_hand(mob/user as mob)
-	playsound(src.loc, "rustle", 50, 1, -5)
+/obj/item/storage/attack_hand(mob/user)
+	playsound(src.loc, "rustle", 50, TRUE, -5)
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:393]: src implied

File: code/game/objects/items/weapons/storage/storage.dm [hunk 11]
```diff
@@ -62,7 +62,7 @@
 			return ..()
 		if(!(src.loc == usr) || (src.loc && src.loc.loc == usr))
 			return
-		playsound(src.loc, "rustle", 50, 1, -5)
+		playsound(src.loc, "rustle", 50, TRUE, -5)
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:65]: src implied

File: code/game/objects/items.dm [hunk 12]
```diff
@@ -624,6 +624,15 @@ GLOBAL_DATUM_INIT(fire_overlay, /image, image("icon" = 'icons/goonstation/effect
 	deltimer(tip_timer) //delete any in-progress timer if the mouse is moved off the item before it finishes
 	closeToolTip(usr)
 
+/obj/item/MouseDrop_T(obj/item/I, mob/user)
+	if(!user || src == I)
+		return
+
+	if(loc && I.loc == loc && istype(loc, /obj/item/storage)) // Are we trying to swap two items in the storage?
+		var/obj/item/storage/S = loc
+		if(S)
```
Comment (farie82) [inline code/game/objects/items.dm:633]: Not required since you checked this on line 631 already

**Comment by Balthusdire** [comment]
This is why I read PRs. Its like crack.

File: code/_onclick/drag_drop.dm [hunk 13]
```diff
@@ -6,11 +6,14 @@
...
 /atom/MouseDrop(atom/over)
-	if(!usr || !over) return
-	if(!Adjacent(usr) || !over.Adjacent(usr)) return // should stop you from dragging through windows
+	if(!usr || !over)
+		return
+	if(!(istype(over, /obj/screen) || (loc && loc == over.loc)))
+		if(!Adjacent(usr) || !over.Adjacent(usr)) // should stop you from dragging through windows
+			return
 
 	spawn(0)
```
Comment (farie82) [inline code/_onclick/drag_drop.dm:15]: Can use `INVOKE_ASYNC` here instead of spawn(0)

File: code/_onclick/drag_drop.dm [hunk 14]
```diff
@@ -6,11 +6,14 @@
...
-		over.MouseDrop_T(src,usr)
+		over.MouseDrop_T(src, usr)
 	return
```
Comment (farie82) [inline code/_onclick/drag_drop.dm:17]: Non required return

File: code/game/objects/items/weapons/storage/storage.dm [hunk 15]
```diff
@@ -197,75 +195,76 @@
...
 	//var/mob/living/carbon/human/H = user
 	var/row_num = 0
-	var/col_count = min(7,storage_slots) -1
+	var/col_count = min(7, storage_slots) - 1
 	if(adjusted_contents > 7)
-		row_num = round((adjusted_contents-1) / 7) // 7 is the maximum allowed width.
-	src.standard_orient_objs(row_num, col_count, numbered_contents)
+		row_num = round((adjusted_contents - 1) / 7) // 7 is the maximum allowed width.
+	standard_orient_objs(row_num, col_count, display_contents)
 	return
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:238]: Non required return

File: code/game/objects/items/weapons/storage/storage.dm [hunk 16]
```diff
@@ -406,17 +406,17 @@
...
+		show_to(user)
 	else
 		..()
 		for(var/mob/M in range(1))
 			if(M.s_active == src)
-				src.close(M)
-	src.add_fingerprint(user)
+				close(M)
+	add_fingerprint(user)
 	return
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:420]: Non required return

Comment (dearmochi) [reply code/game/objects/items/weapons/storage/storage.dm]: All done, cheers for the feedback!

File: code/_onclick/drag_drop.dm [hunk 17]
```diff
@@ -6,13 +6,13 @@
...
+			return
 
-	spawn(0)
-		over.MouseDrop_T(src,usr)
-	return
+	INVOKE_ASYNC(over, .proc/MouseDrop_T, src, usr)
 
 // recieve a mousedrop
 /atom/proc/MouseDrop_T(atom/dropping, mob/user)
-	return
```
Comment (farie82) [inline code/_onclick/drag_drop.dm:18]: This one can stay to show the readers that this is meant to be empty and overridden if needed.

File: code/game/objects/items/weapons/storage/storage.dm [hunk 18]
```diff
@@ -117,63 +115,58 @@
...
-	return
+	closer.screen_loc = "[mx+1],[my]"
 
 //This proc draws out the inventory and places the items on it. It uses the standard position.
-/obj/item/storage/proc/standard_orient_objs(var/rows, var/cols, var/list/obj/item/display_contents)
+/obj/item/storage/proc/standard_orient_objs(rows, cols, list/datum/numbered_display/display_contents)
 	var/cx = 4
 	var/cy = 2+rows
-	src.boxes.screen_loc = "4:16,2:16 to [4+cols]:16,[2+rows]:16"
+	boxes.screen_loc = "4:16,2:16 to [4+cols]:16,[2+rows]:16"
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:169]: spacing here

File: code/game/objects/items/weapons/storage/storage.dm [hunk 19]
```diff
@@ -117,63 +115,58 @@
...
 		O.screen_loc = "[cx],[cy]"
 		O.layer = ABOVE_HUD_LAYER
 		O.plane = ABOVE_HUD_PLANE
 		cx++
 		if(cx > mx)
 			cx = tx
 			cy--
-	src.closer.screen_loc = "[mx+1],[my]"
-	return
+	closer.screen_loc = "[mx+1],[my]"
```
Comment (farie82) [inline code/game/objects/items/weapons/storage/storage.dm:163]: spacing

Comment (dearmochi) [reply code/game/objects/items/weapons/storage/storage.dm:163]: All done

File: code/game/objects/items/weapons/storage/storage.dm [hunk 20]
```diff
@@ -21,10 +21,10 @@
 	var/display_contents_with_number	//Set this to make the storage item group contents of the same type and display them as a number.
 	var/allow_quick_empty	//Set this variable to allow the object to have the 'empty' verb, which dumps all the contents on the floor.
 	var/allow_quick_gather	//Set this variable to allow the object to have the 'toggle mode' verb, which quickly collects all items from a tile.
-	var/collection_mode = 1;  //0 = pick one at a time, 1 = pick all on tile
+	var/collection_mode = TRUE  //FALSE = pick one at a time, TRUE = pick all on tile
```
Comment (AffectedArc07) [inline code/game/objects/items/weapons/storage/storage.dm:24]: If youre going to make this `TRUE/FALSE` can you rename the var to something else?
`collection_mode = TRUE` does not make sense here

File: code/game/objects/items/weapons/storage/storage.dm [hunk 21]
```diff
@@ -117,73 +115,68 @@
...
-	src.boxes.screen_loc = "4:16,2:16 to [4+cols]:16,[2+rows]:16"
+	var/cy = 2 + rows
+	boxes.screen_loc = "4:16,2:16 to [4 + cols]:16,[2 + rows]:16"
 
 	if(display_contents_with_number)
 		for(var/datum/numbered_display/ND in display_contents)
 			ND.sample_object.mouse_opacity = MOUSE_OPACITY_OPAQUE
 			ND.sample_object.screen_loc = "[cx]:16,[cy]:16"
-			ND.sample_object.maptext = "<font color='white'>[(ND.number > 1)? "[ND.number]" : ""]</font>"
+			ND.sample_object.maptext = "<font color='white'>[(ND.number > 1) ? "[ND.number]" : ""]</font>"
```
Comment (AffectedArc07) [inline code/game/objects/items/weapons/storage/storage.dm:175]: Cn you throw `face='Small Fonts'` into here to make the text render crisp

Comment (dearmochi) [reply code/game/objects/items/weapons/storage/storage.dm:175]: All done

**Review by variableundefined — CHANGES_REQUESTED** [review]
Please deconflict

**Pull request merged**